File: SecondYear/semestrul4/MPP/PROIECT/JavaProjectORM/ProjectPersistance/src/main/java/ro/mpp2025/javaprojectui/database/ProbaRepository.java
package ro.mpp2025.javaprojectui.database;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.stereotype.Repository;
import ro.mpp2025.javaprojectui.*;
import ro.mpp2025.javaprojectui.interfaces.IProbaRepository;
import ro.mpp2025.javaprojectui.utils.JdbcUtils;

import java.sql.*;
import java.util.Optional;

@Repository
@Component
public class ProbaRepository extends AbstractRepository<Integer, Proba> implements IProbaRepository {

    private static final Logger logger= LogManager.getLogger(ProbaRepository.class);

    private final Connection connection;

    public ProbaRepository() throws SQLException {
        logger.info("Initializing CarsDBRepository");
        this.connection = JdbcUtils.getInstance().getConnection();
    }

    @Override
    protected PreparedStatement findOneQuery(Integer id) throws SQLException {
        String query = "SELECT * FROM probe " +
                "INNER JOIN categorievarsta on probe.varsta = categorievarsta.id" +
                " WHERE proba_id = ?";
        PreparedStatement statement = connection.prepareStatement(query);
        statement.setObject(1, id);

        return statement;
    }

    @Override
    protected PreparedStatement findAllQuery() throws SQLException {
        String query = "SELECT * FROM probe" +
                " INNER JOIN categorievarsta on probe.varsta = categorievarsta.id";
        return connection.prepareStatement(query);
    }

    @Override
    protected PreparedStatement saveQuery(Proba entity) throws SQLException {
        String query = "INSERT INTO probe(tip, varsta) VALUES (?, ?)";
        PreparedStatement statement = connection.prepareStatement(query);
        statement.setString(1, entity.getTip());
        statement.setObject(2, entity.getVarsta().getId());
        return statement;
    }

    @Override
    protected Proba buildEntity(ResultSet resultSet) throws SQLException {

        Integer id = resultSet.getInt("proba_id");
        String tip = resultSet.getString("tip");
        Integer varstaId = resultSet.getInt("varsta");

        CategorieVarstaRepository cvRepo = new CategorieVarstaRepository();
        Optional<CategorieVarsta> varstaOp =  cvRepo.findOne(varstaId);

        if (varstaOp.isPresent()) {
            CategorieVarsta varsta = varstaOp.get();
            Proba proba = new Proba(tip, varsta);
            proba.setId(id);
            return proba;
        } else {
            throw new RuntimeException("CategorieVarsta not found for id: " + varstaId);
        }

    }

    @Override
    public PreparedStatement getProbaByName(String name) throws SQLException {
        String query = "SELECT * FROM probe WHERE tip = ?";
        PreparedStatement statement = connection.prepareStatement(query);
        statement.setString(1, name);

        return statement;
    }

    @Override
    public PreparedStatement getProbaByNameAndRange(String name, Integer range) throws SQLException {
        String query = "SELECT * FROM probe WHERE tip = ? AND varsta = ?";
        PreparedStatement statement = connection.prepareStatement(query);
        statement.setString(1, name);
        statement.setInt(2, range);

        return statement;
    }

    @Override
    public PreparedStatement updateProba(Proba entity) throws SQLException {
        String query = "UPDATE probe SET tip = ?, varsta = ? WHERE proba_id = ?";
        PreparedStatement statement = connection.prepareStatement(query);
        statement.setString(1, entity.getTip());
        statement.setObject(2, entity.getVarsta().getId());
        statement.setObject(3, entity.getId());
        return statement;
    }

    @Override
    public PreparedStatement deleteProba(Integer id) throws SQLException {
        String query = "DELETE FROM probe WHERE proba_id = ?";
        PreparedStatement statement = connection.prepareStatement(query);
        statement.setObject(1, id);
        return statement;
    }

    @Override
    public Optional<Proba> save(Proba entity) throws SQLException {
        try(PreparedStatement statement = saveQuery(entity)){
            int result = statement.executeUpdate();
            if(result > 0){
                try(Statement stmt = connection.createStatement();
                    ResultSet resultSet = stmt.executeQuery("SELECT last_insert_rowid()");){
                    if(resultSet.next()){
                        entity.setId(resultSet.getInt(1));
                    }
                    else{
                        throw new SQLException("Creating proba failed, no ID obtained.");
                    }
                }
                return Optional.of(entity);
            }
        } catch (Exception e) {
            System.out.println("Eroare DB:" + e.getMessage());
        }
        return Optional.empty();
    }

    public Optional<Proba> update(Proba entity) throws SQLException {
        try (PreparedStatement statement =  updateProba(entity)) {
            int result = statement.executeUpdate();
            logger.info("Updated proba: " + entity);
            if(result > 0 ){
                return Optional.of(entity);
            }
        } catch (SQLException e) {
            logger.error("Error update proba: {}", e.getMessage());
            throw new RuntimeException(e);
        }

        return Optional.empty();
    }

    public void delete(Integer id) throws SQLException {
        try (PreparedStatement statement = deleteProba(id)) {
            int result = statement.executeUpdate();
            logger.info("Deleted proba with id: " + id);
        } catch (SQLException e) {
            logger.error("Error delete proba: {}", e.getMessage());
            throw new RuntimeException(e);
        }
    }
    public Proba findProbaByName(String name) {

        try (PreparedStatement statement =  getProbaByName(name)) {
            try (ResultSet resultSet = statement.executeQuery()) {
                if (resultSet.next()) {
                    Proba entity = buildEntity(resultSet);
                    entity.setId(resultSet.getInt("proba_id"));
                    return entity;
                }
            }
        } catch (SQLException e) {
            logger.error("Error loading entity: {}", e.getMessage());
            throw new RuntimeException(e);
        }

        return null;
    }

    public Proba findProbaByNameAndRange(String name, Integer range) {
        try (PreparedStatement statement =  getProbaByNameAndRange(name, range)) {
            try (ResultSet resultSet = statement.executeQuery()) {
                if (resultSet.next()) {
                    Proba entity = buildEntity(resultSet);
                    entity.setId(resultSet.getInt("proba_id"));
                    return entity;
                }
            }
        } catch (SQLException e) {
            logger.error("Error loading entity: {}", e.getMessage());
            throw new RuntimeException(e);
        }

        return null;
    }


}

File: SecondYear/semestrul4/MPP/PROIECT/CSharpProjectNetworking/Persistence/database/ProbaRepository.cs
using System.Data;
using log4net;
using Model;

namespace Persistence.database;
public class ProbaRepository : AbstractRepository<int, Proba>, IProbaRepo
    {
        private static readonly ILog logger = LogManager.GetLogger(typeof(OrganizatorRepository));
        private readonly IDbConnection connection;

        public ProbaRepository()
        {
            logger.Info("Initializing ProbaRepository...");
            connection = JdbcUtils.GetInstance().GetConnection();
        }

        protected override IDbCommand FindOneQuery(int id)
        {
            string query = "SELECT * FROM probe " +
                           "INNER JOIN categorievarsta on probe.varsta = categorievarsta.cv_id " +
                           "WHERE proba_id = @id";
            var command = connection.CreateCommand();
            command.CommandText = query;

            var param = command.CreateParameter();
            param.ParameterName = "@id";
            param.Value = id;
            command.Parameters.Add(param);

            return command;
        }

        protected override IDbCommand FindAllQuery()
        {
            string query = "SELECT * FROM probe INNER JOIN categorievarsta on probe.varsta = categorievarsta.cv_id";
            var command = connection.CreateCommand();
            command.CommandText = query;
            return command;
        }

        protected override IDbCommand SaveQuery(Proba entity)
        {
            string query = "INSERT INTO probe(tip, varsta) VALUES ( @tip, @varsta)";
            var command = connection.CreateCommand();
            command.CommandText = query;


            var paramTip = command.CreateParameter();
            paramTip.ParameterName = "@tip";
            paramTip.Value = entity.Tip;
            command.Parameters.Add(paramTip);

            var paramVarsta = command.CreateParameter();
            paramVarsta.ParameterName = "@varsta";
            paramVarsta.Value = entity.Varsta.Id;
            command.Parameters.Add(paramVarsta);

            return command;
        }

        protected override Proba BuildEntity(IDataReader reader)
        {
            int varstaID = reader.GetInt32(reader.GetOrdinal("varsta"));
            CategorieVarstaRepository repository = new CategorieVarstaRepository();
            CategorieVarsta varsta = repository.FindOne(varstaID);
            
            int id = reader.GetInt32(reader.GetOrdinal("proba_id"));
            string tip = reader.GetString(reader.GetOrdinal("tip"));

            Proba proba = new Proba(tip, varsta);
            proba.Id = id;

            return proba;
        }
        
        public IDbCommand GetProbaByName(string name)
        {
            // Add the join to categorievarsta table
            string query = "SELECT * FROM probe " +
                           "INNER JOIN categorievarsta on probe.varsta = categorievarsta.cv_id " +
                           "WHERE tip = @name";
            var command = connection.CreateCommand();
            command.CommandText = query;

            var param = command.CreateParameter();
            param.ParameterName = "@name";
            param.Value = name;
            command.Parameters.Add(param);

            return command;
        }

        public IDbCommand GetProbaByNameAndRange(string name, int range)
        {
            // Add the join to categorievarsta table
            string query = "SELECT * FROM probe " +
                           "INNER JOIN categorievarsta on probe.varsta = categorievarsta.cv_id " +
                           "WHERE tip = @name AND varsta = @range";
            var command = connection.CreateCommand();
            command.CommandText = query;

            var paramName = command.CreateParameter();
            paramName.ParameterName = "@name";
            paramName.Value = name;
            command.Parameters.Add(paramName);

            var paramRange = command.CreateParameter();
            paramRange.ParameterName = "@range";
            paramRange.Value = range;
            command.Parameters.Add(paramRange);

            return command;
        }

        public Proba FindProbaByName(string name)
        {
            try
            {
                using var command = GetProbaByName(name);
                using var reader = command.ExecuteReader();
                
                if (reader.Read())
                {
                    Proba entity = BuildEntity(reader);
                    entity.Id = reader.GetInt32(reader.GetOrdinal("proba_id"));
                    return entity;
                }
            }
            catch (Exception e)
            {
                logger.Error($"Error loading entity: {e.Message}");
                throw new Exception(e.Message);
            }

            return null;
        }

        public Proba FindProbaByNameAndRange(string name, int range)
        {
            try
            {
                using var command = GetProbaByNameAndRange(name, range);
                using var reader = command.ExecuteReader();
                
                if (reader.Read())
                {
                    Proba entity = BuildEntity(reader);
                    entity.Id = reader.GetInt32(reader.GetOrdinal("proba_id"));
                    return entity;
                }
            }
            catch (Exception e)
            {
                logger.Error($"Error loading entity: {e.Message}");
                throw new Exception(e.Message);
            }

            return null;
        }
    }